knora/excel2xml.py

```diff
@@ -0,0 +1,2127 @@
+    id: str,
+    permissions: str = "res-default",
+    ark: Optional[str] = None,
+    iri: Optional[str] = None
+) -> etree._Element:
+    """
+    Creates an empty resource element, with the attributes as specified by the arguments
+
+    Args:
+        The arguments correspond 1:1 to the attributes of the <resource> element.
+
+    Returns:
+        The resource element, without any children, but with the attributes
+        ``<resource label=label restype=restype id=id permissions=permissions ark=ark iri=iri></resource>``
+
+    Examples:
+        >>> resource = make_resource(...)
+        >>> resource.append(make_text_prop(...))
+        >>> root.append(resource)
+
+    See https://docs.dasch.swiss/latest/DSP-TOOLS/dsp-tools-xmlupload/#describing-resources-with-the-resource-element
+    """
+
+    kwargs = {
+        "label": label,
+        "restype": restype,
+        "id": id,
+        "permissions": permissions,
+        "nsmap": xml_namespace_map
+    }
+    if ark:
+        kwargs["ark"] = ark
+    if iri:
+        kwargs["iri"] = iri
+    if ark and iri:
+        handle_warnings(f"Both ARK and IRI were provided for resource '{label}' ({id}). The ARK will override the IRI.")
+
+    resource_ = etree.Element(
+        "{%s}resource" % (xml_namespace_map[None]),
+        **kwargs
+    )
+    return resource_
+
+
+def make_bitstream_prop(
+    path: str,
+    permissions: str = "prop-default",
+    calling_resource: str = ""
+) -> etree._Element:
+    """
+    Creates a bitstream element that points to path
+
+    Args:
+        path: path to a valid file that will be uploaded
+        permissions: permissions string
+        calling_resource: the name of the parent resource (for better error messages)
+
+    Examples:
+        >>> resource = make_resource(...)
+        >>> resource.append(make_bitstream_prop("data/images/tree.jpg"))
+        >>> root.append(resource)
+
+    See https://docs.dasch.swiss/latest/DSP-TOOLS/dsp-tools-xmlupload/#bitstream
+    """
+
+    if not os.path.isfile(path):
+        handle_warnings(f"The following is not the path to a valid file: {path} (resource '{calling_resource}')",
+                        stacklevel=3)
+    prop_ = etree.Element("{%s}bitstream" % (xml_namespace_map[None]), permissions=permissions,
+                          nsmap=xml_namespace_map)
+    prop_.text = path
+    return prop_
+
+
+def _format_bool(unformatted: Union[bool, str, int], name: str, calling_resource: str) -> str:
+    if isinstance(unformatted, str):
+        unformatted = unformatted.lower()
+    if unformatted in (False, "false", "0", 0, "no"):
+        return "false"
+    elif unformatted in (True, "true", "1", 1, "yes"):
+        return "true"
+    else:
+        raise BaseError(f"Invalid boolean format for prop '{name}' in resource '{calling_resource}': '{unformatted}'")
+
+
+def make_boolean_prop(
+    name: str,
+    value: Union[PropertyElement, str, int, bool, Iterable[Union[PropertyElement, str, int, bool]]],
```

Comment:
to me, this type signature indicates that the function is too much overloaded, which indicates that the overall design should be improved:  
I would either always or never want this to be expecting a PropertyElement; and I would try to separate the iterable case out if possible at all.  
One option could be that you insist this to require a PropertyElement (and if you don't have one, you need to create one before being able to call this function), and that PropertyElement actually always contains a iterable as value, and if you only have one value, you simply pack it into a list.  
Because what you do here is to create an artificial choke point where you allow very different things to go into one variable by means of union types, just to then split that up again by type checking the same variable. That's not the cleanest design. (it would at least look tidy with the python 3.10 match case, but we shouldn't require 3.10 in DSP-Tools yet.)

Reply:
You're right, I allow too much in one variable. I have now removed the support for Iterables in the 'value' parameter of all make_*_prop() functions.